### PR TITLE
Creates previous_employment_details table

### PR DESCRIPTION
**Pull request by mohinid**
## Notion card
https://www.notion.so/saeloun/Add-tables-for-personal-employment-Allocated-devices-compensation-71ef49d0c38e4242bd2998120141372d

## Summary
1. Created previous_employment_details table with fields: company_name and role.
2. Added associations to EmploymentDetails and CompanyUsers model.
3. Added. validations on fields.
4. Added RSpec
5. Tested manually on Console.

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Manually verified associations and validations on the Rails console.

### Checklist:

- [x] I have manually tested all workflows
- [x] I have performed a self-review of my own code
- [x] I have added automated tests for my code

### Preview of ERD
Employee detail schema ERD

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -135,7 +138,9 @@
     t.bigint "client_id", null: false
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.string "external_view_key"
     t.index ["client_id"], name: "index_invoices_on_client_id"
+    t.index ["external_view_key"], name: "index_invoices_on_external_view_key", unique: true
```
Comment (rohitjoshixyz) [inline db/schema.rb:143]: This looks like an unrelated schema change

**Review by rohitjoshixyz — COMMENTED** [review]
Check schema file, rest LGTM

File: db/schema.rb (hunk 1 above)
Comment (mohinid) [reply db/schema.rb:143]: Yes it took some invoice changes!

**Review by apoorv-mishra — CHANGES_REQUESTED** [review]
Could you resolve conflicts?

**Review by akhilgkrishnan — COMMENTED** [review]
@mohinid I've some concerns about this association. The think the previous company details should be directly associated with the User model.  For a user, the previous employment should same for each company and it does not vary from company to company.
Eg: Consider, Currently a user belongs to 2 companies A and B. Same user has worked for 3 companies before. As per your flow, the user has 2 employment_details ( for company A and Company B) and each employee details have 3 same previous employees details.

**Comment by mohinid** [comment]
> @akhilgkrishnan Well, it makes sense that we associate the model to users since we are storing only `name` and `role` for now. That won't change for any particular individual. It will also reduce a join so let's go for it 👍 @mohinid

Thanks, @rohitjoshixyz, @akhilgkrishnan I will change it.

**Comment by keshavbiswa** [comment]
> , Currently a user belongs to 2 companies A and B. Same user has worked for 3 companies before

Agreed. Shouldn't the same be true for employee_details table?

The user would have the same employee_details regardless of whichever company he is part of.

**Comment by mohinid** [comment]
> > , Currently a user belongs to 2 companies A and B. Same user has worked for 3 companies before
> 
> Agreed. Shouldn't the same be true for employee_details table?
> 
> The user would have the same employee_details regardless of whichever company he is part of.

Hi @keshavbiswa Not really. Please check the fields of employment_details table on ERD. It has employee_id, designation etc which is specific to each company.

**Comment by keshavbiswa** [comment]
> 

So, if current_user has two `company_users` records, this would mean the current_user would have two seperate `employment_details`?

**Comment by keshavbiswa** [comment]
Okay, understood. But if that's the case then, is there a need to create employment_details table? I mean wouldn't it be easier to add these attributes to company_users table directly?

**Comment by mohinid** [comment]
> Okay, understood. But if that's the case then, is there a need to create employment_details table? I mean wouldn't it be easier to add these attributes to company_users table directly?

I think from what I understand a good DB has normalization and having separate table for similar data is a better approach.

**Comment by mohinid** [comment]
Updated reference from CompanyUser to User as discussed. 
Also, tested on local.
Will update the ERD in next PR!

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -16,6 +16,7 @@
     it { is_expected.to have_many(:identities).dependent(:delete_all) }
     it { is_expected.to have_many(:project_members).dependent(:destroy) }
     it { is_expected.to have_many(:timesheet_entries) }
+    it { is_expected.to have_many(:previous_employment_details) }
```
Comment (keshavbiswa) [inline spec/models/user_spec.rb:19]:
```suggestion
    it { is_expected.to have_many(:previous_employment_details).dependent(:destroy }
```

File: app/models/employment_detail.rb [hunk 3]
```diff
@@ -25,6 +25,7 @@
 
 class EmploymentDetail < ApplicationRecord
   belongs_to :company_user
+  has_many :previous_employment_details
```
[the remark this answers is not in the document]
Comment (keshavbiswa) [reply app/models/employment_detail.rb:28]: Also could you add a one-liner test for this association, just like the user_spec?

**Review by keshavbiswa — COMMENTED** [review]
Added few suggestions for missing tests, rest LGTM!

File: app/models/user.rb [hunk 4]
```diff
@@ -62,6 +62,7 @@ class User < ApplicationRecord
   has_many :timesheet_entries
   has_many :identities, dependent: :delete_all
   has_one :wise_account, dependent: :destroy
+  has_many :previous_employment_details, dependent: :destroy
```
Comment (keshavbiswa) [inline app/models/user.rb:65]: Could you also add a spec for this one?

**Review by keshavbiswa — APPROVED** [review]
Thanks for making all the changes. Looks good to me!

**Comment by mohinid** [comment]
> Thanks for making all the changes. Looks good to me!

Thanks a lot @keshavbiswa for your time!! :)

**Review by rohitjoshixyz — APPROVED** [review]
LGTM
